callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java

```diff
@@ -0,0 +1,309 @@
+package io.callstats.sdk;
+
+import io.callstats.sdk.messages.AuthorizeRequest;
+import io.callstats.sdk.messages.AuthorizeResponse;
+import io.callstats.sdk.messages.ChallengeRequest;
+import io.callstats.sdk.messages.ChallengeResponse;
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.http.HttpResponse;
+import org.apache.http.ParseException;
+import org.apache.http.util.EntityUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import com.google.gson.Gson;
+import com.ning.http.client.Response;
+
+public class CallStatsAuthenticator {
+
+	private static final Logger logger = LogManager.getLogger("CallStatsAuthenticator");
+	private static final int RESPONSE_STATUS_SUCCESS = 200;
+	private static final int SERVER_ERROR = 500;
+	private static final int INVALID_PROTO_FORMAT_ERROR = 400;
+	private static final int INVALID_PARAM_ERROR = 403;
+	private static final int GATEWAY_ERROR = 502;
+	
+	private static final String challengeUrl = "/o/challenge";
+	private static final String authorizeUrl = "/o/authorize";
+	
+	private CallStatsInitListener listener;
+	private int  authenticationRetryTimeout = 5000;
+	private static final ScheduledExecutorService scheduler = 
+			  Executors.newSingleThreadScheduledExecutor();
+	private Gson gson;
+	
+	private String token;
+	private String expires;
+	
+	public String getToken() {
+		return token;
+	}
+	public void setToken(String token) {
+		this.token = token;
+	}
+	public String getExpires() {
+		return expires;
+	}
+	public void setExpires(String expires) {
+		this.expires = expires;
+	}	
+
+	public CallStatsAuthenticator(CallStatsInitListener listener) {
+		this.listener = listener;
+		gson = new Gson();
+	}
+	
+	public void doAuthentication(final int appId,final String appSecret,final String bridgeId,final CallStatsHttpClient httpClient) {
```

Comment:
I wonder if the interface looks better if we put these dependencies into the constructor of the CallStatsAuthenticator and then just call 'doAuthentication()' to kick off the authentication process?